pkg/trace/api/api.go

```diff
@@ -354,6 +375,43 @@ func (r *HTTPReceiver) logStats() {
 	}
 }
 
+func (r *HTTPReceiver) watchdog(now time.Time) {
+	wi := watchdog.Info{
+		Mem: watchdog.Mem(),
+		CPU: watchdog.CPU(now),
+	}
+
+	rate, err := sampler.CalcPreSampleRate(r.conf.MaxCPU, wi.CPU.UserAvg, r.PreSampler.RealRate())
+	if err != nil {
+		log.Warnf("problem computing pre-sample rate: %v", err)
+	}
+
+	r.PreSampler.SetRate(rate)
+	r.PreSampler.SetError(err)
+
+	stats := r.PreSampler.Stats()
+
+	info.UpdatePreSampler(*stats)
+	info.UpdateWatchdogInfo(wi)
+
+	metrics.Gauge("datadog.trace_agent.heap_alloc", float64(wi.Mem.Alloc), nil, 1)
+	metrics.Gauge("datadog.trace_agent.cpu_percent", wi.CPU.UserAvg*100, nil, 1)
+	metrics.Gauge("datadog.trace_agent.presampler_rate", stats.Rate, nil, 1)
+
+	if float64(wi.Mem.Alloc) > r.conf.MaxMemory && r.conf.MaxMemory > 0 {
+		log.Warn("memory exceeds threshold (apm_config.max_memory), requests will be rate-limited")
+		if atomic.SwapInt64(&r.refuse, 1) != 0 {
+			// we're still not accepting requests, do a garbage collection;,
+			// potentially blocking the program here is the least of our problems
+			runtime.GC()
```

Comment:
After the GC, do you want to re-evaluate the memory allocated to see if we are below -> if we can accept payloads again right away?

Reply:
Please see my comment here: https://github.com/DataDog/datadog-agent/pull/3239/files#r277622152. Are you ok with this?